src/js/tech/setup-sourceset.js

```diff
@@ -42,42 +161,175 @@ const setupSourceset = function(tech) {
+  const el = tech.el();
+
+  if (el.firstSourceWatch_) {
+    return;
+  }
+
+  el.firstSourceWatch_ = true;
+  const oldAppend = el.append;
+  const oldAppendChild = el.appendChild;
+  const oldInsertAdjacentHTML = el.insertAdjacentHTML;
+  const innerDescriptor = getInnerHTMLDescriptor(el);
+
+  el.appendChild = function() {
+    const retval = oldAppendChild.apply(el, arguments);
+
+    sourcesetLoad(tech);
+
+    return retval;
+  };
+
+  if (oldAppend) {
+    el.append = function() {
+      const retval = oldAppend.apply(el, arguments);
+
+      sourcesetLoad(tech);
+
+      return retval;
+    };
+  }
+
+  if (oldInsertAdjacentHTML) {
+    el.insertAdjacentHTML = function() {
+      const retval = oldInsertAdjacentHTML.apply(el, arguments);
+
+      sourcesetLoad(tech);
+
+      return retval;
+    };
+  }
+
+  Object.defineProperty(el, 'innerHTML', {
+    get: innerDescriptor.get.bind(el),
+    set(v) {
+      const retval = innerDescriptor.set.call(el, v);
+
+      sourcesetLoad(tech);
+
+      return retval;
+    },
+    configurable: true,
+    enumerable: innerDescriptor.enumerable
+  });
+
+  // on the first sourceset, we need to revert
+  // our changes
+  tech.one('sourceset', (e) => {
+    el.firstSourceWatch_ = false;
+    el.appendChild = oldAppendChild;
+    if (oldAppend) {
+      el.append = oldAppend;
+    }
+    if (oldInsertAdjacentHTML) {
+      el.insertAdjacentHTML = oldInsertAdjacentHTML;
+    }
+
+    Object.defineProperty(el, 'innerHTML', {
+      get: innerDescriptor.get.bind(el),
+      set: innerDescriptor.set.bind(el),
+      configurable: true,
+      enumerable: innerDescriptor.enumerable
+    });
+  });
+};
+
+/**
+ * setup `sourceset` handling on the `Html5` tech. This function
+ * patches the following element properties/functions:
+ *
+ * - `src` - to determine when `src` is set
+ * - `setAttribute()` - to determine when `src` is set
+ * - `load()` - this re-triggers the source selection algorithm, and can
+ *              cause a sourceset.
+ *
+ * If there is no source when we are adding `sourceset` support or during a `load()`
+ * we also patch the functions listed in `firstSourceWatch`.
+ *
+ * @param {Html5} tech
+ *        The tech to patch
+ */
+const setupSourceset = function(tech) {
+  if (!tech.featuresSourceset) {
+    return;
+  }
+
+  const el = tech.el();
+  const srcDescriptor = getSrcDescriptor(el);
+  const oldSetAttribute = el.setAttribute;
+  const oldLoad = el.load;
+
+  // we need to fire sourceset when the player is ready
+  // if we find that the media element had a src when it was
+  // given to us and that tech element is not in a stalled state
+  if (el.src || el.currentSrc && el.initNetworkState_ !== 3) {
+    tech.triggerSourceset(el.src || el.currentSrc);
+  }
+
+  // for some reason adding a source element when a mediaElement has no source
+  // calls `load` internally right away. We need to handle that.
+  if (!el.src && !el.currentSrc && !tech.$$('source').length) {
```

Comment:
OK so I tried this and I do think that it would make sense, but it has a bad interaction with:
https://github.com/videojs/video.js/blob/378d98ee87b6ac48536d0b8f263bee1f76c68aab/src/js/tech/html5.js#L49

and I don't think we want to change that code right now. So I will be keeping everything as is and open another pr after this one to make this change.

Reply:
making it in another PR sounds OK.